### PR TITLE
[Feature] Add R support

### LINKED ISSUE #953: [Feature]: Add R support
I was following the updates and just want to thank all for the great project!

Would love to see language support for R. I got it partially work following the suggestions from Nvim-lspconfig but there are many details that I could not figure out (newbie in lua and programming in general). 
Thanks in advance.

### PR DESCRIPTION
**Pull request by abzcoding**
# Description

`R` language support including formatting

Fixes #953

please run the following to install dependencies:
```bash
R -e 'install.packages("languageserver",repos = "http://cran.us.r-project.org")'
R -e 'install.packages("formatR",repos = "http://cran.us.r-project.org")'
R -e 'install.packages("readr",repos = "http://cran.us.r-project.org")'
```

### CODE REVIEW

**Comment by ChristianChiarulli** [comment]
that comment should probably be included on top of the method somewhere for reference.

**Comment by JulianTao** [comment]
Thanks for the quick actions, @abzcoding !

It works for the .r files but not perfectly for the .rmd files. 

**If I open a .rmd file, the r_language_server will not be active**; but somehow if I open a .r file first, the r_language_server becomes active; then if I open a .rmd file, the server is now active. 

**Another issue is that the formatter does not work for the rmd files.** In a simple r code chunk, I left some space before the right parenthesis, lsp detected the format issue.  But if I use :Format or <Leader>lf, errors occur. Running :messages gives the follow:

> Formatter: failed to run formatter R. Error in parse(text = lines, keep.source = TRUE) :   attempt to use zero-length variable nameCalls: <Anonymous> ... mask_comments -> parse_data -> <Anonymous> -> parse_source
-> parseExecution halted

**Comment by abzcoding** [comment]
@JulianTao 

:facepalm: yeah that is my bad gonna fix it for `rmd` files, so are the `r` files working fine?

**Comment by JulianTao** [comment]
@abzcoding ,  `r` files are working fine!

**Comment by abzcoding** [comment]
> @abzcoding , `r` files are working fine!

I've added a single file, can you pull it and check to see if fixes the issue or not?

**Comment by JulianTao** [comment]
> > @abzcoding , `r` files are working fine!
> 
> I've added a single file, can you pull it and check to see if fixes the issue or not?

Lsp is now active. Formatter still does not work. Same message

**Comment by abzcoding** [comment]
> > > @abzcoding , `r` files are working fine!
> > 
> > 
> > I've added a single file, can you pull it and check to see if fixes the issue or not?
> 
> Lsp is now active. Formatter still does not work. Same message

can you provide a sample `rmd` file so I can test it locally :pleading_face:

**Comment by JulianTao** [comment]
I could not upload an `rmd` file directly. This is an `md` file.  Save it as `test.Rmd`. Hope it works.
[test.md](https://github.com/ChristianChiarulli/LunarVim/files/6825442/test.md)
.

**Comment by abzcoding** [comment]
@JulianTao 
:sob: [FormatR](https://github.com/yihui/formatR) doesn't support `rmd` formatting [yet](https://github.com/yihui/formatR/issues/72)

do you any any formatter that supports `rmd` files?

**Comment by JulianTao** [comment]
> @JulianTao
>  [FormatR](https://github.com/yihui/formatR) doesn't support `rmd` formatting [yet](https://github.com/yihui/formatR/issues/72)
> 
> do you know any formatter that supports `rmd` files?

I do not know other formatters that support `rmd` files. There is a workaround for the output files, though. [Reformat R chunks](https://bookdown.org/yihui/rmarkdown-cookbook/opts-tidy.html). This option helps tidy the format in the output file, but not in the `rmd` file itself. I guess it is good enough for my use. 

I think the issue  #953 is fixed. Thank you so much!

**Comment by abzcoding** [comment]
> > @JulianTao
> > [FormatR](https://github.com/yihui/formatR) doesn't support `rmd` formatting [yet](https://github.com/yihui/formatR/issues/72)
> > do you know any formatter that supports `rmd` files?
> 
> I do not know other formatters that support `rmd` files. There is a workaround for the output files, though. [Reformat R chunks](https://bookdown.org/yihui/rmarkdown-cookbook/opts-tidy.html). This option helps tidy the format in the output file, but not in the `rmd` file itself. I guess it is good enough for my use.
> 
> I think the issue #953 is fixed. Thank you so much!

np mate , hopefully, the support for `rmd` files gonna be added in the near future

File: lua/lang/r.lua [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  }
+end
+
+M.format = function()
+  O.formatters.filetype["r"] = {
+    function()
+      return {
+        exe = O.lang.r.formatter.exe,
+        args = O.lang.r.formatter.args,
+        stdin = not (O.lang.r.formatter.stdin ~= nil),
```
Comment (Tastyep) [inline lua/lang/r.lua:25]: This is equivalent to stdin = (O.lang.r.formatter.stdin == nil), right ?
So stdin == true if O.lang.r.formatter.stdin == nil, is this right ?

Comment (abzcoding) [reply lua/lang/r.lua:25]: 🤦  yeah you are right, gotta fix them all in a new PR probably 😭

**Comment by jolars** [comment]
> > @JulianTao
> > [FormatR](https://github.com/yihui/formatR) doesn't support `rmd` formatting [yet](https://github.com/yihui/formatR/issues/72)
> > do you know any formatter that supports `rmd` files?
> 
> I do not know other formatters that support `rmd` files. There is a workaround for the output files, though. [Reformat R chunks](https://bookdown.org/yihui/rmarkdown-cookbook/opts-tidy.html). This option helps tidy the format in the output file, but not in the `rmd` file itself. I guess it is good enough for my use.
> 
> I think the issue #953 is fixed. Thank you so much!

The [styler package](https://github.com/r-lib/styler) support Rmarkdown files, just so you know.

**Pull request merged**